dipy/tracking/tests/test_streamline.py

```diff
@@ -791,25 +805,103 @@ def test_orient_by_rois():
+    new_streamlines = orient_by_rois(streamlines,
+                                     mask1_vol,
+                                     mask2_vol,
+                                     out=None,  # This should return a copy
+                                     affine=None,
+                                     as_generator=False)
+    npt.assert_equal(new_streamlines, flipped_sl)
+    npt.assert_(new_streamlines is not streamlines)
+
+    # Test with affine:
+    x_flipped_sl = [s + affine[:3, 3] for s in flipped_sl]
+    new_streamlines = orient_by_rois(x_streamlines,
+                                     mask1_vol,
+                                     mask2_vol,
+                                     out=None,  # This should return a copy
+                                     affine=affine,
+                                     as_generator=False)
+    npt.assert_equal(new_streamlines, x_flipped_sl)
+    npt.assert_(new_streamlines is not x_streamlines)
+
+    # Test providing coord ROIs instead of vol ROIs:
+    new_streamlines = orient_by_rois(x_streamlines,
+                                     mask1_coords,
+                                     mask2_coords,
+                                     out=None,  # This should return a copy
+                                     affine=affine,
+                                     as_generator=False)
+    npt.assert_equal(new_streamlines, x_flipped_sl)
+
+    # Test with as_generator set to True
+    new_streamlines = orient_by_rois(streamlines,
+                                     mask1_vol,
+                                     mask2_vol,
+                                     out=None,
+                                     affine=None,
+                                     as_generator=True)
+
+    npt.assert_(isinstance(new_streamlines, types.GeneratorType))
+    ll = list(new_streamlines)
+    npt.assert_equal(ll, flipped_sl)
+
+    # Test with as_generator set to True and with the affine
+    new_streamlines = orient_by_rois(x_streamlines,
+                                     mask1_vol,
+                                     mask2_vol,
+                                     out=None,
+                                     affine=affine,
+                                     as_generator=True)
+
+    npt.assert_(isinstance(new_streamlines, types.GeneratorType))
+    ll = list(new_streamlines)
+    npt.assert_equal(ll, x_flipped_sl)
+
+    # Test with generator input:
+    new_streamlines = orient_by_rois(generate_sl(streamlines),
+                                     mask1_vol,
+                                     mask2_vol,
+                                     out=None,
+                                     affine=None,
+                                     as_generator=True)
+
+    npt.assert_(isinstance(new_streamlines, types.GeneratorType))
+    ll = list(new_streamlines)
+    npt.assert_equal(ll, flipped_sl)
+
+    # Generator input with an `out` set to something raises an error:
```

Comment:
Generator input with an `out` set to something should **not** raise an error. However, `as_generator=True` and `out` set to something should raise an error. I think that is what you are already doing in the code anyway.